podkrepi-bg/api
# [1/2]: Overhaul of  donations module

- Create a new table _payments_, which will hold information to payment itself such as extPaymentIntentId, chargedAmount, amount, billingName, billingEmail etc..
- Refactored _donations_ table to hold only information regarding the donation itself, such as an amount of payment that was allocated to specific campaign, whether that donation is anonymous, whether there is a wish to that donation.
- Created two functions in Vault module, which would allow to increment/decrement many vaults in a single query.
- Adapted existing API endpoints regarding donations, to the new database schema
- Adapted tests to the new database schema

## Motivation and context
Currently two major features are queued:
- Donation to a whole category
- Importing donations coming from Benevity

It was discussed among the PM, and dev team, that our current structure is not fit for situations, where funds from one payment can be allocated to more than one campaign. 

For this reason a new structure was proposed, which makes the relation between a payment and campaign vaults M:N, with _donations_ table being the "middle-man", connection a payment to campaign vault. Figure below"
![donation-overall-structure](https://github.com/podkrepi-bg/api/assets/27433690/0a8cd895-a32b-4ad8-a4cd-4cbf5378ddd4)

To make this happen, the current _donations_ structure, was split into two tables _payments_ and _donations_, with relation between those tables being 1:N(one payment, can be linked to one or more donations)
![new-donation-structure-FINAL](https://github.com/podkrepi-bg/api/assets/27433690/aeee56ae-6cf7-42bb-ac75-e9a06ff777fd)









## Testing
1. Tested whether the current donation records, are safely migrated to the new structure.
2. Tested payment done via Stripe.
3. Tested payment through affiliateCode
4. Tested payment through IRIS import

### Steps to test
1. Before migration, save the current donation records(Open prisma studio->Donations -> Select one row  initially-> Click on the square to select all rows then CTRL+C) in a file.
2. Run the migration script
3. Payments and Donations count should be the exact same
4. Compare the new payment table, with records saved in step1(e.g. match of extPaymentintentId, billingEmail, chargedAmount etc..)
5. Compare the new donations table, with the records saved in step1(e.g. donation type, amount, targetvaultId, personId etc..)
6. Check whether donation via stripe is successfully imported, in the new structure
7. Check whether donation via IRIS bank import is successfully imported in the new structure.
8. Check whether donation via affiliateCode, is successfully imported in the new structure.

### New endpoints
*/donations/payments* - Get information just for the made payments
